slack/slack_test.go

```diff
@@ -85,6 +87,94 @@ func TestRealSlack(t *testing.T) {
 	})
 }
 
+func TestDebounce(t *testing.T) {
+	Convey("Given a slacker", t, func() {
+		s, messageChan, dfunc := startMockSlackAndCreateSlack()
+		defer dfunc()
+
+		Convey("You can create a debounce tracker", func() {
+			testMessageSuffix := "test messages"
+
+			debounce := 500 * time.Millisecond
+
+			dt := NewDebounceTracker(s, debounce, testMessageSuffix)
+
+			Convey("Which only sends messages after the debounce timeout", func() {
+				dt.SendDebounceMsg(1)
+
+				expectedOutput := fmt.Sprintf("%s1 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+
+				dt.SendDebounceMsg(2)
+				So(checkNoMessage(messageChan), ShouldBeTrue)
+
+				<-time.After(debounce)
+
+				expectedOutput = fmt.Sprintf("%s2 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+			})
+
+			Convey("Which only sends unique messages", func() {
+				dt.SendDebounceMsg(1)
+
+				expectedOutput := fmt.Sprintf("%s1 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+
+				dt.SendDebounceMsg(1)
+				So(checkNoMessage(messageChan), ShouldBeTrue)
+
+				<-time.After(debounce)
+
+				So(checkNoMessage(messageChan), ShouldBeTrue)
+			})
+
+			Convey("Which sends the highest number seen in the debounce period", func() {
+				dt.SendDebounceMsg(4)
+
+				expectedOutput := fmt.Sprintf("%s4 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+
+				dt.SendDebounceMsg(3)
+				dt.SendDebounceMsg(0)
+				dt.SendDebounceMsg(6)
+				dt.SendDebounceMsg(5)
+
+				<-time.After(debounce)
+
+				expectedOutput = fmt.Sprintf("%s6 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+			})
+
+			Convey("Which always sends the final 0 message", func() {
+				dt.SendDebounceMsg(1)
+
+				expectedOutput := fmt.Sprintf("%s1 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+
+				dt.SendDebounceMsg(3)
+
+				<-time.After(debounce)
+
+				expectedOutput = fmt.Sprintf("%s3 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+
+				dt.SendDebounceMsg(2)
+				dt.SendDebounceMsg(0)
+
+				<-time.After(debounce)
+
+				expectedOutput = fmt.Sprintf("%s2 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+
+				<-time.After(debounce)
+
+				expectedOutput = fmt.Sprintf("%s0 %s", BoxPrefixInfo, testMessageSuffix)
+				checkMessage(expectedOutput, messageChan)
+			})
+		})
```

Comment:
There should be a test that it doesn't panic with a nil slacker, and any other exceptions that have already been guarded against in the code.